core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java

```diff
@@ -0,0 +1,348 @@
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InMemoryFileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/**
+ * Catalog implementation that uses in-memory data-structures to store the namespaces and tables.
+ * This class doesn't touch external resources and can be utilized to write unit tests without side
+ * effects. It uses {@link InMemoryFileIO}.
+ */
+public class InMemoryCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Closeable {
+  private static final Joiner SLASH = Joiner.on("/");
+  private static final Joiner DOT = Joiner.on(".");
+
+  private final ConcurrentMap<Namespace, Map<String, String>> namespaceDb;
+  private final ConcurrentMap<TableIdentifier, String> tableDb;
+  private FileIO io;
+  private String catalogName;
+  private String warehouseLocation;
+
+  public InMemoryCatalog() {
+    namespaceDb = Maps.newConcurrentMap();
+    tableDb = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public String name() {
+    return catalogName;
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    this.catalogName = name != null ? name : InMemoryCatalog.class.getSimpleName();
+
+    String warehouse = properties.getOrDefault(CatalogProperties.WAREHOUSE_LOCATION, "");
+    this.warehouseLocation = warehouse.replaceAll("/*$", "");
+    this.io = new InMemoryFileIO();
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    return new InMemoryTableOperations(io, tableIdentifier);
+  }
+
+  @Override
+  protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
+    return SLASH.join(
+        defaultNamespaceLocation(tableIdentifier.namespace()), tableIdentifier.name());
```

Comment:
Did you intend to include the warehouse location? It doesn't matter much since all IO is in memory, but the warehouse location is pulled out of properties and is available.